File: benchmarks/tsb/bench_clip_with_bounds.ts
/**
 * Benchmark: clipSeriesWithBounds / clipDataFrameWithBounds — per-element clipping with Series/array bounds.
 * Outputs JSON: {"function": "clip_with_bounds", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, clipSeriesWithBounds, clipDataFrameWithBounds } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 5;
const ITERATIONS = 20;

const data = Float64Array.from({ length: ROWS }, (_, i) => (i % 200) - 100);
const lowerArr = Array.from({ length: ROWS }, () => -30);
const upperArr = Array.from({ length: ROWS }, () => 30);
const s = new Series(data);
const lowerSeries = new Series(lowerArr);
const upperSeries = new Series(upperArr);

const dfCols: Record<string, number[]> = {};
for (let c = 0; c < 4; c++) {
  dfCols[`col${c}`] = Array.from({ length: ROWS }, (_, i) => (i + c * 10) % 200 - 100);
}
const df = new DataFrame(dfCols);
const dfLower = new Series(Array.from({ length: ROWS }, () => -30));
const dfUpper = new Series(Array.from({ length: ROWS }, () => 30));

for (let i = 0; i < WARMUP; i++) {
  clipSeriesWithBounds(s, { lower: lowerSeries, upper: upperSeries });
  clipDataFrameWithBounds(df, { lower: dfLower, upper: dfUpper, axis: 0 });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  clipSeriesWithBounds(s, { lower: lowerSeries, upper: upperSeries });
  clipDataFrameWithBounds(df, { lower: dfLower, upper: dfUpper, axis: 0 });
}
const total = performance.now() - start;

console.log(JSON.stringify({ function: "clip_with_bounds", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }));

File: benchmarks/pandas/bench_clip_with_bounds.py
"""
Benchmark: pandas Series.clip / DataFrame.clip with array/Series bounds.
Outputs JSON: {"function": "clip_with_bounds", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

ROWS = 100_000
WARMUP = 5
ITERATIONS = 20

data = np.array([(i % 200) - 100 for i in range(ROWS)], dtype=float)
lower_arr = np.full(ROWS, -30.0)
upper_arr = np.full(ROWS, 30.0)
s = pd.Series(data)
lower_s = pd.Series(lower_arr)
upper_s = pd.Series(upper_arr)

df_cols = {f"col{c}": [(i + c * 10) % 200 - 100 for i in range(ROWS)] for c in range(4)}
df = pd.DataFrame(df_cols, dtype=float)
df_lower = pd.Series(np.full(ROWS, -30.0))
df_upper = pd.Series(np.full(ROWS, 30.0))

for _ in range(WARMUP):
    s.clip(lower=lower_s, upper=upper_s)
    df.clip(lower=df_lower, upper=df_upper, axis=0)

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.clip(lower=lower_s, upper=upper_s)
    df.clip(lower=df_lower, upper=df_upper, axis=0)
total = (time.perf_counter() - start) * 1000

print(json.dumps({"function": "clip_with_bounds", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))
